lib/optimizer/src/Transformer/AmpBoilerplate.php

```diff
@@ -68,31 +67,44 @@ public function transform(Document $document, ErrorCollection $errors)
     }
 
     /**
-     * Remove all <style> and <noscript> tags except for the <style amp-custom> tag.
+     * Remove all <style> and <noscript> tags which are for the boilerplate.
      *
      * @param Document $document Document to remove the tags from.
      */
     private function removeStyleAndNoscriptTags(Document $document)
     {
-        $headNode = $document->head->firstChild;
-        $devMode = DevMode::isActiveForDocument($document);
-        while ($headNode) {
-            $nextSibling = $headNode->nextSibling;
-            if ($headNode instanceof DOMElement && (! $devMode || ! DevMode::hasExemptionForNode($headNode))) {
-                switch ($headNode->tagName) {
-                    case Tag::STYLE:
-                        if (! $headNode->hasAttribute(Attribute::AMP_CUSTOM)) {
-                            $document->head->removeChild($headNode);
-                        }
-                        break;
-                    case Tag::NOSCRIPT:
-                        $document->head->removeChild($headNode);
-                        break;
-                }
+        /**
+         * Style element.
+         *
+         * @var DOMElement $style
+         */
+        foreach (iterator_to_array($document->head->getElementsByTagName('style')) as $style) {
+            if (! $this->hasBoilerplateAttribute($style)) {
+                continue;
             }
+            if (Tag::NOSCRIPT === $style->parentNode->nodeName) {
+                $style->parentNode->parentNode->removeChild($style->parentNode);
+            } else {
+                $style->parentNode->removeChild($style);
+            }
+        }
+    }
 
-            $headNode = $nextSibling;
+    /**
+     * Check whether an element has a boilerplate attribute.
+     *
+     * @param DOMElement $element Element to check if it .
+     * @return bool Whether the element has a boilerplate attribute.
+     */
+    private function hasBoilerplateAttribute(DOMElement $element)
```

Comment:
We already have a `hasNoBoilerplateAttribute()` method, which does something completely unrelated, so this could be confusing. I'd suggest renaming this.
```suggestion
    private function isBoilerplateStyle(DOMElement $element)
```